qiskit/dagcircuit/dagcircuit.py

```diff
@@ -1371,6 +1371,36 @@ def substitute_node(self, node, op, inplace=False):
             self._decrement_op(node.op)
         return new_node
 
+    def swap_nodes(self, node1, node2):
+        """Swap connected nodes e.g. due to commutation.
+
+        Args:
+            node1 (OpNode): predecessor node
+            node2 (OpNode): successor node
+
+        Raises:
+            DAGCircuitError: if either node is not an OpNode
+        """
+        if not (isinstance(node1, DAGOpNode) and isinstance(node2, DAGOpNode)):
+            raise DAGCircuitError("nodes to swap are not both OpNodes")
+        try:
+            connected_edges = self._multi_graph.get_all_edge_data(node1._node_id, node2._node_id)
+        except rx.NoEdgeBetweenNodes:
+            # since nodes aren't connected there's nothing to do
+            return
```

Comment:
This feels like it ought to be some sort of exception - we're being asked to swap nodes that can't be swapped.  I don't think silently doing nothing is safe.

Reply:
The case I was thinking of for not raising, which this doesn't quite cover, is was a situation like,
```
q_0: ──■───────
       │       
q_1: ──┼────■──
     ┌─┴─┐  │  
q_2: ┤ X ├──┼──
     └───┘┌─┴─┐
q_3: ─────┤ X ├
          └───┘
```
where the method might be looping through some list of many swap candidates. When it encounters this one it can merrily  continue without having to do anything. I'm not sure of an easy way, however, to catch when operations can't be considered to be swapped due to other intervening instructions. Maybe `DAGDependency` would work better for such a case. I'll raise here.